src/plugins/intel_gpu/src/runtime/kernels_cache.cpp

```diff
@@ -89,7 +89,9 @@ size_t kernels_cache::get_max_kernels_per_batch() const {
     GPU_DEBUG_IF(debug_config->max_kernels_per_batch >= 1) {
         return static_cast<size_t>(debug_config->max_kernels_per_batch);
     }
-    return 8;
+    // [TODO] We found a performance issue when multiple kernels are built as a single program.
+    // The below WA code needs to be restored after it is fixed.
+    return _config.get_property(ov::intel_gpu::allow_new_shape_infer) ? 4 : 8;
```

Comment:
Basically allow_new_shape_infer is not for checking model's dynamicness. So I'd like to suggest to set this value by checking dynamicness of the program.

Reply:
I updated it as you reviewed. Thank you!